[ConstraintSystem] Change the order in which we visit disjunctions.

rudkx opened the pull request:
Attempt to visit disjunctions that are associated with applies where
we have at least some useful information about the types of all of the
arguments before visiting other disjunctions.

Two tests here got faster, and one slightly slower. One of the
faster tests is actually moving from test/ to the slow/ directory in
validation-test because despite going from 16s to less than 1s, it was
still borderline for what we consider the slow threshold, so I made
the test more complex. The one that got a little slower is
rdar22022980, which I also made more complex so that it is clearly
"slow" by the way we are testing it.

slower:
  rdar22022980.swift

faster:
  rdar33688063.swift
  expression_too_complex_4.swift

Commit:
[ConstraintSystem] Change the order in which we visit disjunctions.

Attempt to visit disjunctions that are associated with applies where
we have at least some useful information about the types of all of the
arguments before visiting other disjunctions.

Two tests here got faster, and one slightly slower. One of the
faster tests is actually moving from test/ to the slow/ directory in
validation-test because despite going from 16s to less than 1s, it was
still borderline for what we consider the slow threshold, so I made
the test more complex. The one that got a little slower is
rdar22022980, which I also made more complex so that it is clearly
"slow" by the way we are testing it.

slower:
  rdar22022980.swift

faster:
  rdar33688063.swift
  expression_too_complex_4.swift

rudkx commented:
@swift-ci Please smoke test

rudkx commented:
@swift-ci Please test source compatibility

xedin commented on lib/Sema/CSSolver.cpp:
```diff
@@ -1358,6 +1358,224 @@ static Constraint *selectBestBindingDisjunction(
+    llvm::SetVector<Constraint *> constraints;
+    cs.getConstraintGraph().gatherConstraints(
+        rep, constraints, ConstraintGraph::GatheringKind::EquivalenceClass,
+        [](Constraint *constraint) {
```
Nit: default callback already does `return true;` so you don't have to explicitly provide it...

Commit:
[ConstraintSystem] Remove lambda that returns true since this argument
is defualted in the same way.

xedin commented on lib/Sema/CSSolver.cpp:
```diff
@@ -1358,6 +1358,224 @@ static Constraint *selectBestBindingDisjunction(
+// in the given function type.
+static bool haveTypeInformationForAllArguments(AnyFunctionType *fnType,
+                                               ConstraintSystem &cs) {
+  for (auto param : fnType->getParams())
```
This really looks like a great candidate for `llvm::all_of`...

rudkx replied:
Yup, I never went back and cleaned that up after moving the logic into a separate function. I'll push changes to do so.

xedin commented on lib/Sema/CSSolver.cpp:
```diff
@@ -1358,6 +1358,221 @@ static Constraint *selectBestBindingDisjunction(
+  cs.getConstraintGraph().gatherConstraints(
+      rep, disjunctions, ConstraintGraph::GatheringKind::EquivalenceClass,
+      [](Constraint *match) {
+        return match->getKind() == ConstraintKind::Disjunction;
```
It looks like logic from the `for` loop at the end of this method could be moved into filtering callback, because that test doesn't really require any additional information besides what disjunction has...

rudkx replied:
Yeah this normally returns zero or one matches, but you're right I can just fold the logic in. I'll push a change for that as well.

xedin reviewed (approved):
LGTM! I have couple of nit comments which could be addressed separately. Reminds me of https://github.com/apple/swift/pull/11723

slavapestov commented on lib/Sema/CSSolver.cpp:
```diff
@@ -1358,6 +1358,221 @@ static Constraint *selectBestBindingDisjunction(
+static bool haveTypeInformationForAllArguments(AnyFunctionType *fnType,
+                                               ConstraintSystem &cs) {
+  for (auto param : fnType->getParams())
+    if (!havePotentialTypesOrLiteralConformances(param.getType(), cs))
```
Can you use getPlainType() here? getType() wraps the result in an InOutType if its an inout parameter, but is otherwise equivalent. If you need to special-case inouts, just check param.isInOut() instead.

Thanks for using FunctionType::getParams() instead of getInput() though, I appreciate the help in transitioning over to the new representation :-)

xedin replied:
+1, sorry I keep on forgetting that `getType()` does that... :(

Commit:
[ConstraintSystem] Use llvm::all_of.

Commit:
[ConstraintSystem] Fold some logic into the gatherConstraints predicate.

Commit:
[ConstraintSystem] Use Param::getPlainType() rather than Param::getType().

rudkx commented:
@swift-ci Please smoke test

rudkx commented:
@swift-ci Please smoke test Linux platform

rudkx commented:
@swift-ci Please test source compatibility

Commit:
[ConstraintSystem] Only use the new disjunction ordering under -swift-version 5.

This will minimize the chance of breaking code. Currently SwiftLint
has one "too complex" expression with this change. Further changes to
the solver may improve that situation, and potentially allow us to
move this out from -swift-version 5 if we're willing to take the risk
of breaking some code as a result.

rudkx commented:
@swift-ci Please smoke test

Commit:
Merge branch 'master' into fix-ordering

rudkx commented:
@swift-ci Please smoke test

rudkx commented:
@swift-ci Please test source compatibility
